add rename

dfries opened the pull request:
This avoids creating the target directory if the old and new path are
identical.

required for js-git data integrity fix I'm about to push.

dfries committed:
add rename

For git objects and references, fs-db can now write to a temporary
file, and when it is ready 'rename' it into place, which avoids
exposing partially written files or truncating a file if the program
exits at the wrong time.  This avoids creating the target directory if
the old and new path are identical.

creationix commented on lib/node-fs.js:
```diff
@@ -115,6 +116,25 @@ function writeFile(path, binary, callback) {
+  if (oldBase !== newBase) {
+    mkdirp(nodePath.dirname(path), function (err) {
+      if (err) {
+        callback(err);
```
Missing return.  Should be `return callback(err);`

creationix replied:
Though I guess since you're using an else, it's not strictly required, it's just the style I use everywhere.

creationix commented on lib/node-fs.js:
```diff
@@ -115,6 +116,25 @@ function writeFile(path, binary, callback) {
+      }
+    });
+  } else {
+    nodeFs.rename(oldPath, newPath, callback);
```
If you move this case to the top on the `if` and use a return, then the else block won't be needed and the larger section can save a level of indentation.

dfries committed:
code style/callback cleanup

Making this consistent with the style in other modules, and for good
reason mkdirp was missing a return on a callback which would fall
out to make a second callback.

dfries commented:
The commit fc738bb4a801e9a258e4c50ae6bf49ae16eb998b is ready to merge, as the original style of my rename was identical to the preceding writeFile I went ahead and made the changes you were suggesting to update the style in the rest of the file, and caught a missing return on a callback at the end of the file in the process.

creationix commented:
Thanks
